### PR TITLE
[dockers] install nvidia-dali-cudaXXX

### PR DESCRIPTION
**Pull request by ydcjeff**
## What does this PR do?

Fixes dali installation for #3721 

## Before submitting
- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.

## PR review
 - [x] Is this pull request ready for review? (if not, please submit in draft mode)

Anyone in the community is free to review the PR once the tests have passed.    
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Review by SeanNaren — APPROVED** [review]

File: dockers/base-conda/Dockerfile [hunk 1]
```diff
@@ -30,6 +30,7 @@ ARG PYTHON_VERSION=3.7
 ARG PYTORCH_VERSION=1.6
 ARG PYTORCH_CHANNEL=pytorch
 ARG CONDA_VERSION=4.7.12
+ARG NVIDIA_DALI_VERSION=100
```
Comment (Borda) [inline dockers/base-conda/Dockerfile:33]:
```suggestion
```

File: dockers/base-conda/Dockerfile [hunk 2]
```diff
@@ -104,6 +105,7 @@ RUN \
     # Install remaining requirements
     pip install -r requirements-extra.txt --upgrade-strategy only-if-needed && \
     pip install -r requirements-test.txt --upgrade-strategy only-if-needed && \
+    pip install --extra-index-url https://developer.download.nvidia.com/compute/redist nvidia-dali-cuda${NVIDIA_DALI_VERSION} && \
```
Comment (Borda) [inline dockers/base-conda/Dockerfile:108]:
```suggestion
    pip install --extra-index-url https://developer.download.nvidia.com/compute/redist nvidia-dali-cuda${CUDA_VERSION/./} && \
```

File: dockers/base-cuda/Dockerfile [hunk 3]
```diff
@@ -28,6 +28,7 @@ FROM nvidia/cuda:${CUDA_VERSION}-cudnn${CUDNN_VERSION}-devel-ubuntu18.04
 
 ARG PYTHON_VERSION=3.7
 ARG PYTORCH_VERSION=1.6
+ARG NVIDIA_DALI_VERSION=100
```
Comment (Borda) [inline dockers/base-cuda/Dockerfile:31]:
```suggestion
```

File: dockers/base-cuda/Dockerfile [hunk 4]
```diff
@@ -93,6 +94,7 @@ RUN \
 
     # Install all requirements
     pip install -r requirements/devel.txt --upgrade-strategy only-if-needed --use-feature=2020-resolver && \
+    pip install --extra-index-url https://developer.download.nvidia.com/compute/redist nvidia-dali-cuda${NVIDIA_DALI_VERSION} && \
```
Comment (Borda) [inline dockers/base-cuda/Dockerfile:97]:
```suggestion
    pip install --extra-index-url https://developer.download.nvidia.com/compute/redist nvidia-dali-cuda${CUDA_VERSION/./} && \
```

**Review by Borda — COMMENTED** [review]
can we install it from source rather?

**Comment by Borda** [comment]
ok, I have tried to install specific versions according to CUDA version, but it seems to fail as they have 100 or 110 only...
then I am nor sure about compatibility as we use CUDA 10.1 or 10.2...
lets check installation from source

**Comment by SeanNaren** [comment]
> ok, I have tried to install specific versions according to CUDA version, but it seems to fail as they have 100 or 110 only...
> then I am nor sure about compatibility as we use CUDA 10.1 or 10.2...
> lets check installation from source

@Borda I've tested their cuda 100 version with cuda 10.2, does it make a difference? I think dali-cuda100 should work for both

**Comment by Borda** [comment]
> @Borda I've tested their cuda 100 version with cuda 10.2, does it make a difference? I think dali-cuda100 should work for both

I would use the compiled one in Drone config and here rather build for a specific version as we also do for APEx now
https://docs.nvidia.com/deeplearning/dali/user-guide/docs/compilation.html

**Comment by SeanNaren** [comment]
Ok cool, let's see how this works!

**Review by awaelchli — APPROVED** [review]

**Comment by Borda** [comment]
@ydcjeff I got it up to some strange errors with parsing, mind have a look?
```
CMake Error at cmake/Utils.cmake:345 (string):
  string sub-command REPLACE requires at least four arguments.
Call Stack (most recent call first):
  dali/operators/python_function/CMakeLists.txt:19 (build_per_python_lib)

CMake Error at cmake/Utils.cmake:351 (string):
  string sub-command REPLACE requires at least four arguments.
Call Stack (most recent call first):
  dali/operators/python_function/CMakeLists.txt:19 (build_per_python_lib)

CMake Error at cmake/Utils.cmake:352 (string):
  string sub-command REPLACE requires at least four arguments.
Call Stack (most recent call first):
  dali/operators/python_function/CMakeLists.txt:19 (build_per_python_lib)

-- Adding dependencies to target `dali`: '/usr/local/cuda/targets/x86_64-linux/lib/libnvjpeg_static.a;/usr/local/cuda/targets/x86_64-linux/lib/libnppicc_static.a;/usr/local/cuda/targets/x86_64-linux/lib/libnppc_static.a;/usr/local/cuda/targets/x86_64-linux/lib/libculibos.a;opencv_core;opencv_imgproc;opencv_imgcodecs;/usr/lib/x86_64-linux-gnu/libjpeg.so;/usr/lib/x86_64-linux-gnu/libtiff.so;/usr/lib/x86_64-linux-gnu/libsndfile.so;avformat;avformat;avcodec;avfilter;avutil;ffts;CUTLASS;cocoapi;/usr/lib/x86_64-linux-gnu/libprotobuf.a;/usr/local/cuda/targets/x86_64-linux/lib/libcudart_static.a;rt;pthread;m;dl'
CMake Error at cmake/Utils.cmake:345 (string):
  string sub-command REPLACE requires at least four arguments.
Call Stack (most recent call first):
  dali/python/CMakeLists.txt:19 (build_per_python_lib)

CMake Error at cmake/Utils.cmake:351 (string):
  string sub-command REPLACE requires at least four arguments.
Call Stack (most recent call first):
  dali/python/CMakeLists.txt:19 (build_per_python_lib)

CMake Error at cmake/Utils.cmake:352 (string):
  string sub-command REPLACE requires at least four arguments.
Call Stack (most recent call first):
  dali/python/CMakeLists.txt:19 (build_per_python_lib)
```

**Comment by ydcjeff** [comment]
@Borda @SeanNaren does dali only run with gpu, right?

**Comment by Borda** [comment]
I spend some time yesterday a few hours trying to compile it... 
I would give it a little more to try - fix the cmake issue above or open issue on DALI project
the ultimate solution would be using the binary and hope that it be compatible with the particular CUDAs

**Comment by SeanNaren** [comment]
I agree with binaries, I even think that this is something that should be rigorously tested as it is a third party library required for users. I wouldn't spend too much time thinking about building from source and just either moving with the binary, or just skipping the test.

**Comment by Borda** [comment]
@SeanNaren @ydcjeff reverted to binaries :]

**Review by SeanNaren — APPROVED** [review]

**Review by Borda — APPROVED** [review]

**Pull request merged**